cirq/linalg/decompositions.py

```diff
@@ -1045,3 +1046,34 @@ def _gamma(u: np.ndarray) -> np.ndarray:
         u @ yy @ u.T @ yy, where yy = Y ⊗ Y
     """
     return u @ YY @ u.T @ YY
+
+
+def extract_right_diag(u: np.ndarray, atol=1e-15):
+    """Extract a diagonal unitary from a 3-CNOT two-qubit unitary.
+
+    Returns a 2-CNOT unitary D that is diagonal, so that U @ D needs only
+    two CNOT gates in case the original unitary is a 3-CNOT unitary.
+
+    See Proposition V.2 in Minimal Universal Two-Qubit CNOT-based Circuits.
+    https://arxiv.org/abs/quant-ph/0308033
+
+    Args:
+        u: three-CNOT two-qubit unitary
+        atol: the absolute tolerance for avoiding division by zero
+    Returns:
+        diagonal extracted from U
+    """
+    t = _gamma(transformations.to_special(u).T).T.diagonal()
+    k = np.real(t[0] + t[3] - t[1] - t[2])
+    if np.abs(k) < atol:
```

Comment:
I think you shouldn't need this conditional (and the atol parameter). Working around this singularity is exactly why people came up with [arctan2](https://en.wikipedia.org/wiki/Atan2)! This should work:

```python
psi = np.arctan2(np.imag(np.sum(t)), k)
```

(i.e. remove the conditional, unindent arctan below, replace arctan with arctan2 and "/ k" with ", k"). Note that [numpy's arctan2](https://numpy.org/doc/stable/reference/generated/numpy.arctan2.html) is also well-behaved when both arguments are zero (how cool is that?!).

Reply:
Dude, that is so cool, very nice, I didn't know about that!!! :)